Upgrade README.md documentation,basic compilation usages (#389)
quarkus-qe/quarkus-test-suite: README.md

# Quarkus Test Suite

> The aim of this test suite is to validate Quarkus framework features along the time. Under the hood we are using [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) that we designed for this proposal. 

Reviewer: Not the main aim of this test suite. Can be omitted IMO. I guess you wanted to reference another chapter https://github.com/quarkus-qe/quarkus-test-suite#test-framework which mentioned in https://github.com/quarkus-qe/quarkus-test-suite/issues/118

The test suite includes:
- Bare metal scenario testing
- OpenShift scenario testing
- Native testing

## How-to run Quarkus test suite

Please before start be sure that you have all [prerequisites](#Prerequisites) installed.

### Openshift

For running the tests in OpenShift, it is expected that the user is logged into an OpenShift project:
- with OpenShift 4, run `oc login https://api.<cluster name>.<domain>:6443`

To verify that you're logged into correct project, you can run `oc whoami` and `oc project`.

> By default [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) use [ephemeral namespaces](https://github.com/quarkus-qe/quarkus-test-framework/wiki/6.-Openshift#disable-ephemeral-namespaces) in order to be deterministic and don't populate you environment with random objects. After each scenarios all the objects that are required, will be removed.

Reviewer: ```suggestion
By default [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) uses [ephemeral namespaces](https://github.com/quarkus-qe/quarkus-test-framework/wiki/6.-Openshift#disable-ephemeral-namespaces) in order to be deterministic and don't populate your OCP environment with random resources. After each scenario, all the instantiated resources will be removed.
```

Practical Example:

User: `I would like to run http-minimum module.` 

```shell

> mvn clean verify -Dall-modules -Dopenshift -pl http/http-minimum 

Reviewer: Explain here why we need enable 2 profiles at the same time

```

### Openshift / Native

Please read [Openshift](#Openshift) section first and login into OCP.

When we are running a [native compilation](https://quarkus.io/guides/building-native-image) the flow is the same as the regular way, the only difference is that we need to compile our application first with GraalVM/Mandrel in order to generate the binary application. To do that we will add the flag `native` to our maven command.   

Practical Example:

> User: `I would like to run http-minimum module.`

Reviewer: ```suggestion
User: `I would like to run http-minimum module in native mode.`
```

> Practical Example:
> 
> User: `I would like to run http-minimum module.`
> 
> ```shell
> mvn clean verify -Dall-modules -Dnative -Dopenshift -pl http/http-minimum 
> ```

Reviewer: ```suggestion
**Practical Example:**

User: `I would like to run http-minimum module.`

```shell
mvn clean verify -Dall-modules -Dnative -Dopenshift -pl http/http-minimum 
```
```
Explain why we need to enable two profiles at the same time

Quarkus test framework will reuse the Native binary generated by Maven to run the test, except if the scenario provides a build property, then it will generate a new Native binary.

More info about how to generate native images could be found in Quarkus [building-native-image](https://quarkus.io/guides/building-native-image) website, and customize your native image generation by adding some [configuration flags](https://quarkus.io/guides/building-native-image#configuration-reference) to your maven command.

> For example, 

Reviewer: ```suggestion
Practical Example:
```

User: `I want to use a custom GraalVM Docker image and 5GB of memory`

```shell
mvn clean verify -Dall-modules -Dnative -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-native-image:21.3-java11 -Dquarkus.native.native-image-xmx=5g -Dopenshift -pl http/http-minimum 
```

> Or what about use the GraalVM that I have installed in my local environment instead of use a Docker image with GraalVM

Reviewer: These two use cases: local graal && customized docker graal should be more emphasized.
Two subchapters of ### Openshift / Native can be introduced for that. And move the content to them

Be sure that GraalVM is installed by running the following command otherwise you will need to install it
`gu --version`
`gu install native-image`

And then type this example

User: `I want to use my local GraalVM in order to build my binary application`

```shell
mvn clean verify -Dall-modules -Dquarkus.package.type=native -Dopenshift -pl http/http-minimum 
```

### Bare metal

Firstly be sure that docker/podman is running

```shell
docker run hello-world
```

> Then just run a maven clean verify command pointing to your scenario. 

Reviewer: Remove this line

Practical Example:

User: `I would like to run http-minimum module.`

```shell
mvn clean verify -Dall-modules -pl http/http-minimum 
```

#### Bare metal / Native

Same as [Openshift/native](#Openshift-/-Native) scenarios, Quarkus test framework will reuse the Native binary generated by Maven to run your tests.

Practical Example:

> User: `I would like to run http-minimum module.`

Reviewer: ```suggestion
User: `I would like to run http-minimum module in native mode.`
```

```shell
mvn clean verify -Dall-modules -Dnative -pl http/http-minimum 
```

All the above example for openshift are also valid for Bare metal, just remove the flag `-Dopenshift` and play with [native image generation properties](https://quarkus.io/guides/building-native-image#configuration-reference)

### Additional notes

I recommend you to have a look the main `pom.xml` and pay attention to some useful topics as how the scenarios are categorized by topics or profiles, or some global properties as `quarkus.platform.version` that could be overwritten by a flag. 

> For example

Reviewer: ```suggestion
Practical Example:
```

As a user I would like to run all core modules of Quarkus `2.2.3.Final`

```shell
mvn clean verify -Dquarkus-core-only -Dquarkus.platform.version=2.2.3.Final 

> ```

Reviewer: I don't understand this sentence :/
Does it applies to QuarkusTestSuite? There is no `quarkus-core-only` in the main pom.xml

# Quarkus Test Suite

The aim of this test suite is to validate Quarkus framework features along the time. Under the hood we are using [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) that we designed for this proposal. 

The test suite includes:
- Bare metal scenario testing
- OpenShift scenario testing
- Native testing

## How-to run Quarkus test suite

> Please before start be sure that you have all [prerequisites](#Prerequisites) installed.

Reviewer 2: This is not good for the good first impression

We can assume bare metal use case as the first option, so I would cover `mvn clean verify`

next to it cover how to run native mode in bare metal - `mvn clean verify -Dnative` ?

third item would be how to run concrete test in jvm mode in baremetal

after that you can add line like "Refer to the section below for additional detail"

Reviewer 2: The impression needs to be "Nice, it's easy to use"

Author: I've re-worked this section. Let me know how do you see now!

### Openshift

For running the tests in OpenShift, it is expected that the user is logged into an OpenShift project:
- with OpenShift 4, run `oc login https://api.<cluster name>.<domain>:6443`

To verify that you're logged into correct project, you can run `oc whoami` and `oc project`.

By default [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) uses [ephemeral namespaces](https://github.com/quarkus-qe/quarkus-test-framework/wiki/6.-Openshift#disable-ephemeral-namespaces) in order to be deterministic and don't populate your OCP environment with random resources. After each scenario, all the instantiated resources will be removed.

Practical Example:

> User: `I would like to run http-minimum module.` 

Reviewer 2: in OpenShift

```shell
mvn clean verify -Dall-modules -Dopenshift -pl http/http-minimum 
```

**Note:** here we are combining two profiles, `openshift` in order to trigger Openshift execution mode and `all-modules` to enable `http/http-minimum` that is located under `http-modules` or also `all-modules` profile

> 

Reviewer 2: It would be good to add a note to check profiles defined in pom.xml to see the key testing profiles.

Author: I have added a new section called `Profiles overview`, WDYT?

# Quarkus Test Suite

Test Suite for Quarkus integration scenarios. Among the standard Quarkus test features, scenarios hardly rely on [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) features

The test suite includes:
- Bare metal scenario testing
- OpenShift scenario testing
- Native testing

## How-to run Quarkus test suite

Please before start be sure that you have all [prerequisites](#Prerequisites) installed.

> As a general rule if you have Docker, Java11 and maven installed in your local environment the only command that you need to run is:

Reviewer 2: Pls replace lines 14 - 19 with this text:

Docker, JDK 11+, and Apache Maven 3.8 are the base requirements to tun the test suite using the following command:
```
mvn clen verify
```
Refer to [prerequisites](#Prerequisites) section for more details about the tooling.

Maven commands can be executed from the root directory or from the sub-modules. Additionally, if you want to run a single scenario you can use `-Dit.test=<TEST_CLASS_NAME>` flag.

> `mvn clen verify`

Reviewer: ```suggestion
`mvn clean verify`
```

You can run it over the main `pom.xml` if you want to run the whole test suite or into a sub-project level if you want to run a single module.
Additionally, if you want to run a single scenario of a single module you could use `-Dit.test=<TEST_CLASS_NAME>` flag, for example:

Let's say that we want to only test `HttpMinimumIT` located over `http/http-minimum` module

```shell
cd http/http-minimum
mvn clean verify -Dit.test=HttpMinimumIT
```
And now let's say that you want to run a single test of a scenario that has several ones. For example in module `http/http-advanced` there is a scenario called `HttpAdvancedIT` that has several test, so let's only run `httpServer` test

```shell
cd http/http-advanced
mvn clean verify -Dit.test=HttpAdvancedIT#httpServer
```

To conclude this section only point that you can also add the flag `-Pnative` to the module examples in order to run a native test scenario, for example

```shell
cd http/http-advanced
mvn clean verify -Pnative -Dit.test=HttpAdvancedIT#httpServer
```

> But if you want to know more about how to deploy and run the test suite in Openshift or how to filter modules, overwrite global properties as Quarkus platform version, or customize your native compilations please continue reading the following subsections.

Reviewer 2: OpenShift

### Profiles overview

If you have a look the main `pom.xml` you will notice that there are several profiles or in other words the test suite is a maven monorepo that you can compile and verify at once or by topics. Let's review the main profiles:

* root-modules: this profile talks about Quarkus "core stuff" as configuration or properties. Is a basic stuff that should work as a pre-requisite to other modules.
* http-modules: talks about HTTP extensions and no-application endpoints like `/q/health`
* security-modules: cover all security stuff like oauth, JWT, OpenId, keycloak etc
* messaging-modules: focus in brokers as Kafka or Artemis-AMQP
* monitor-modules: talk about metrics and tracing
* sql-db-modules: is focus in SQL world, panache, hibernate, raw SQL ...
* nosql-db-modules: is focus in noSQL or schemaless DBs as mongoDB

> * spring-modules: is focus un Spring world

Reviewer: start with: _focus on_

* quarkus-cli-tests: enable Quarkus command client test

By default, all your test are running on bare metal or JVM mode, but you can add the following profiles to you maven command in order to swapp to other platforms

* native: turn you compilation into native compilation

> * openshift: turn your test into an openshift deployment/test

Reviewer 2: OpenShift

* redhat-registry: use Redhat docker registry with official supported images instead of community images
* serverless: enable Knative or serverless scenarios
* operator-scenarios: enable operator scenarios, where the ecosytem of your test is going to be deployed by k8s/OCP operators

All of these profiles are not mutual exclusive, indeed we encourage you to combine these profiles in order to run complex scenarios.

For example, 

> As a user I would like to run in Openshift a native version of root, security and SQL modules and also run knative scenarios of those modules.

Reviewer 2: OpenShift

```shell
mvn clean verify -P openshift,serverless,native,root-modules,security-modules,sql-db-modules
```
The above valid statement could be also re-written to the following query:

```shell
mvn clean verify -Dopenshift -Dserverless -Dnative -P root-modules,security-modules,sql-db-modules
```

> That in my opinion is clear because you separate the test suite modules where the scenarios are located to the platform specific configuration (What vs Where/how), but this is just a matter of taste doesn't have any impact. 

Reviewer: don't write in the first person

Note that in maven a profile could be activated by `-P` or by a system property `-D`. Indeed, all our profiles has a common system property called `all-modules`, this property could very useful because allow you to reference all modules at once

For example

```shell
mvn clean verify -Dall-modules -Dnative
```

or run a single module without known in which profile is really located, is like a convenience shortcut!.

```shell
mvn clean verify -Dall-modules -pl http/http-minimum
```

Maybe start talking about tricks is too advance for now, but that I want to show you is the flexibility and feel free to be "bad" and use your imagination to combine all of these profiles in order to model your scenarios. 

> Well if you want to know more about Openshift scenarios, native compilation or fancy tricks, you can continue reading the following sections.

Reviewer 2: OpenShift

> ### Openshift

Reviewer 2: OpenShift

For running the tests in OpenShift, it is expected that the user is logged into an OpenShift project:
- with OpenShift 4, run `oc login https://api.<cluster name>.<domain>:6443`

To verify that you're logged into correct project, you can run `oc whoami` and `oc project`.

By default [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) uses [ephemeral namespaces](https://github.com/quarkus-qe/quarkus-test-framework/wiki/6.-Openshift#disable-ephemeral-namespaces) in order to be deterministic and don't populate your OCP environment with random resources. After each scenario, all the instantiated resources will be removed.

Practical Example:

> User: `I would like to run http-minimum module in Openshift.` 

Reviewer 2: OpenShift

plus all the other places bellow

```shell
mvn clean verify -Dall-modules -Dopenshift -pl http/http-minimum 
```

> **Note:** here we are combining two profiles, `openshift` in order to trigger Openshift execution mode and `all-modules` to enable `http/http-minimum` that is located under `http-modules` or `all-modules` profile

Reviewer 2: OpenShift
